File: api/Dtos/City.cs
using api.Utility.Paging;
using System.ComponentModel.DataAnnotations;

namespace api.Dtos
{
    public class CityRes
    {
        public int CityId { get; set; }
        public string? Name { get; set; }
        public int StateId { get; set; }
    }
    public class CityResWithAreasCount : CityRes
    {
        public int AreasCount { get; set; }
    }

    public class CityResDetails : CityRes
    {
        public int AreasCount { get; set; }
        public string? StateName { get; set; }
        public int CountryId { get; set; }
        public string? CountryName { get; set; }
    }

    public class CityReqEdit
    {
        [Required, MaxLength(255)]
        public string? Name { get; set; }

        // Foreign keys
        public int StateId { get; set; }
    }

    public class CityReqSearch : PagedRequestDto
    {
        public int StateId { get; set; }
    }
}

File: web/src/dtos/City.ts
import Common from "../utility/Common";
import { PagedReq } from "./PagedReq";
import { StateRes } from "./State";

export interface CityRes {
  cityId?: number;
  name?: string;
  stateId?: number;

  state?: StateRes;
}

export interface CityResWithAreasCount extends CityRes {
  areasCount?: number;
}

export interface CityResDetails extends CityRes {
  areasCount?: number;
  stateName?: string;
  countryId?: number;
  countryName?: string;
}

export class CityReqEdit {
  name?: string = "";
  stateId?: number = 0;
  constructor(stateId?: number) {
    this.stateId = stateId;
  }
}

export class CityReqSearch extends PagedReq {
  stateId?: number;

  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy,
      searchText = "",
    }: PagedReq,
    stateId = 0
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.stateId = stateId;
  }
}